src/GraphQL/Internal/Validation.hs

```diff
@@ -1,104 +1,416 @@
+data Selection spread
+  = SelectionField (Field spread)
+  | SelectionFragmentSpread spread
+  | SelectionInlineFragment (InlineFragment spread)
+  deriving (Eq, Show)
+
+-- | A field in a selection set, which itself might have children which might
+-- have fragment spreads.
+data Field spread
+  = Field (Maybe Alias) Name Arguments Directives [Selection spread]
+  deriving (Eq, Show)
 
-{-
-enum DogCommand { SIT, DOWN, HEEL }
+-- | A fragment spread that has a valid set of directives, but may or may not
+-- refer to a fragment that actually exists.
+data UnresolvedFragmentSpread
+  = UnresolvedFragmentSpread Name Directives
+  deriving (Eq, Show)
+
+-- | A fragment spread that refers to fragments which are known to exist.
+data FragmentSpread
+  = FragmentSpread Name Directives (FragmentDefinition FragmentSpread)
+  deriving (Eq, Show)
+
+-- | An inline fragment, which itself can contain fragment spreads.
+data InlineFragment spread
+  = InlineFragment TypeCondition Directives [Selection spread]
+  deriving (Eq, Show)
+
+-- | Traverse through every fragment spread in a selection.
+--
+-- The given function @f@ is applied to each fragment spread. The rest of the
+-- selection remains unchanged.
+--
+-- Note that this is essentially a definition of 'Traversable' for
+-- 'Selection'. However, we probably also want to have other kinds of
+-- traversals (e.g. for transforming values), so best not to bless one kind
+-- with a type class.
+traverseFragmentSpreads :: Applicative f => (a -> f b) -> Selection a -> f (Selection b)
+traverseFragmentSpreads f selection =
+  case selection of
+    SelectionField (Field alias name args directives ss) ->
+      SelectionField <$> (Field alias name args directives <$> childSegments ss)
+    SelectionFragmentSpread x -> SelectionFragmentSpread <$> f x
+    SelectionInlineFragment (InlineFragment typeCond directives ss) ->
+      SelectionInlineFragment <$> (InlineFragment typeCond directives <$> childSegments ss)
+  where
+    childSegments = traverse (traverseFragmentSpreads f)
+
+-- | Ensure a selection has valid arguments and directives.
+validateSelection :: AST.Selection -> Validation (Selection UnresolvedFragmentSpread)
+validateSelection selection =
+  case selection of
+    AST.SelectionField (AST.Field alias name args directives ss) ->
+      SelectionField <$> (Field alias name <$> validateArguments args <*> validateDirectives directives <*> childSegments ss)
+    AST.SelectionFragmentSpread (AST.FragmentSpread name directives) ->
+      SelectionFragmentSpread <$> (UnresolvedFragmentSpread name <$> validateDirectives directives)
+    AST.SelectionInlineFragment (AST.InlineFragment typeCond directives ss) ->
+      SelectionInlineFragment <$> (InlineFragment typeCond <$> validateDirectives directives <*> childSegments ss)
+  where
+    childSegments = traverse validateSelection
+
+-- | Resolve the fragment references in a selection, accumulating a set of
+-- visited fragment names.
```

Comment:
We traverse through the tree of fragment references, "visiting" each and then resolving them. It's language borrowed from the spec, which I also found confusing in context.

Have changed the comment to:

```haskell
-- | Resolve the fragment references in a selection, accumulating a set of
-- the fragment names that we have resolved.
--
-- We're doing a standard depth-first traversal of fragment references, where
-- references are by name, so the set of names can be thought of as a record
-- of visited references.
```